compiler/util/exprAnalysis.cpp

```diff
@@ -54,6 +54,33 @@ bool SafeExprAnalysis::exprHasNoSideEffects(Expr* e) {
         safeExprCache[e] = false;
         return false;
       }
+      else if (exprToMove != NULL) {
+        //
+        // Exposed by AST pattern like this:
+        //            |------- `exprToMove`
+        // (move T (+ A B))
+        // (move A B) -------- `ce`
+        // (move B T)
+        //
+        // Without this check could turn into:
+        //
+        // (move A B)
+        // (move B (+ A B))
+        //
+        // Which is incorrect.
+        //
+        if (ce->isPrimitive(PRIM_MOVE)) {
+          INT_ASSERT(isSymExpr(ce->get(1)));
+          std::vector<SymExpr*> syms;
+          collectSymExprs(exprToMove, syms);
+          for_vector(SymExpr, s, syms) {
+            if (s->var == toSymExpr(ce->get(1))->var) {
+              safeExprCache[e] = false;
+              return false;
+            }
+          }
+        }
```

Comment:
couldn't this also happen through a reference? Or have the variable set with PRIM_ASSIGN?
I generally think this is the right place for the check, I'm just not sure it covers all the possible cases.